snap/component.go

```diff
@@ -126,14 +130,88 @@ func (c *componentPlaceInfo) MountDescription() string {
 	return fmt.Sprintf("Mount unit for %s, revision %s", c.ContainerName(), c.compRevision)
 }
 
-// ReadComponentInfoFromContainer reads ComponentInfo from a snap component container.
-func ReadComponentInfoFromContainer(compf Container) (*ComponentInfo, error) {
+// ReadComponentInfoFromContainer reads ComponentInfo from a snap component
+// container. If snapInfo is not nil, it is used to complete the ComponentInfo
+// information about the component's implicit and explicit hooks, and their
+// associated plugs.
+func ReadComponentInfoFromContainer(compf Container, snapInfo *Info) (*ComponentInfo, error) {
 	yamlData, err := compf.ReadFile("meta/component.yaml")
 	if err != nil {
 		return nil, err
 	}
 
-	return InfoFromComponentYaml(yamlData)
+	componentInfo, err := InfoFromComponentYaml(yamlData)
+	if err != nil {
+		return nil, err
+	}
+
+	// if snapInfo is nil, then we can't complete the component info with
+	// implicit and explicit hooks, so we return the component info as is.
+	//
+	// we could technically create the hooks, but would be unable to bind plugs
+	// to them, so it is probably best to just leave them out.
+	if snapInfo == nil {
+		return componentInfo, nil
+	}
+
+	componentName := componentInfo.Component.ComponentName
+
+	component, ok := snapInfo.Components[componentName]
+	if !ok {
+		return nil, fmt.Errorf("internal error: %q is not a component for snap %q", componentName, snapInfo.RealName)
+	}
+
+	// attach the explicit hooks, these are defined in the snap.yaml. plugs are
+	// already bound to the hooks.
+	componentInfo.Hooks = component.ExplicitHooks
+
+	// attach the implicit hooks, these are not defined in the snap.yaml.
+	// unscoped plugs are bound to the implicit hooks here.
+	addAndBindImplicitComponentHooksFromContainer(compf, componentInfo, snapInfo, component)
+
+	return componentInfo, nil
+}
+
+func addAndBindImplicitComponentHooksFromContainer(compf Container, componentInfo *ComponentInfo, info *Info, component *Component) {
+	hooks, err := compf.ListDir("meta/hooks")
+	if err != nil {
+		return
+	}
+
+	for _, hook := range hooks {
+		addAndBindImplicitComponentHook(componentInfo, info, component, hook)
+	}
+}
+
+func addAndBindImplicitComponentHook(componentInfo *ComponentInfo, snapInfo *Info, component *Component, hook string) {
+	// TODO: ignore unsupported implicit component hooks, or return an error?
```

Comment:
We ignore them for snaps right now. Alfonso mentioned that it might be because of some backwards compatibility, and we might not want to ignore unsupported hooks for components?